mlflow/transformers.py

```diff
@@ -0,0 +1,851 @@
+
+
+def _load_component(root_path: pathlib.Path, component_key: str, component_type):
+    """
+    Loads an individual component object from local disk.
+    """
+    import transformers
+
+    components_dir = root_path.joinpath(_COMPONENTS_BINARY_KEY)
+    component_path = components_dir.joinpath(component_key)
+    component_instance = getattr(transformers, component_type)
+    return component_instance.from_pretrained(component_path)
+
+
+def _generate_base_flavor_configuration(
+    model,
+    task: str,
+) -> Dict[str, str]:
+    """
+    Generates the base flavor metadata needed for reconstructing a pipeline from saved
+    components. This is important because the ``Pipeline`` class does not have a loader
+    functionality. The serialization of a Pipeline saves the model, configurations, and
+    metadata for ``FeatureExtractor``s, ``Processor``s, and ``Tokenizer``s exclusively.
+    This function extracts key information from the submitted model object so that the precise
+    instance types can be loaded correctly.
+    """
+
+    _validate_transformers_task_type(task)
+
+    flavor_configuration = {
+        _TASK_KEY: task,
+        _LOGGED_TYPE_KEY: _get_instance_type(model, True),
+        _INSTANCE_TYPE_KEY: _get_instance_type(model, False),
+        _MODEL_PATH_OR_NAME_KEY: _get_base_model_architecture(model),
+        _PIPELINE_MODEL_TYPE_KEY: _get_instance_type(model.model, False),
+    }
+
+    return flavor_configuration
+
+
+def _get_or_infer_task_type(model, task: Optional[str] = None) -> str:
+    """
+    Validates that a supplied task type is supported by the ``transformers`` library if supplied,
+    else, if not supplied, infers the appropriate task type based on the model type.
+    """
+    if task:
+        _validate_transformers_task_type(task)
+    else:
+        task = _infer_transformers_task_type(model)
+    return task
+
+
+def _infer_transformers_task_type(model) -> str:
+    """
+    Performs inference of the task type, used in generating a pipeline object based on the
+    underlying model's intended use case. This utility relies on the definitions within the
+    transformers pipeline construction utility functions.
+
+    :param model: Either the model or the Pipeline object that the task will be extracted or
+                  inferred from
+    :return: The task type string
+    """
+    from transformers import Pipeline
+    from transformers.pipelines import get_task
+
+    if isinstance(model, Pipeline):
+        return model.task
+    elif isinstance(model, dict):
+        # transformers provides guard conditions for potentially invalid entries
+        return get_task(model[_MODEL_KEY].name_or_path)
+    else:
+        raise MlflowException(
+            f"The provided model type: {type(model)} is not supported. "
+            "Supported model types are: PreTrainedModel or Pipeline.",
+            error_code=BAD_REQUEST,
+        )
```

Comment:
I think this error message makes sense for the function in which it's defined but is confusing when surfaced as part of a user workflow.

For example, I ran the following bad code where model is neither a pipeline nor a dict:

```
# Model references

# dalle-mega
# DALLE_MODEL = "dalle-mini/dalle-mini/mega-1-fp16:latest"  # can be wandb artifact or 🤗 Hub or local folder or google bucket
DALLE_COMMIT_ID = None

# if the notebook crashes too often you can use dalle-mini instead by uncommenting below line
DALLE_MODEL = "dalle-mini/dalle-mini/mini-1:v0"

# VQGAN model
VQGAN_REPO = "dalle-mini/vqgan_imagenet_f16_16384"
VQGAN_COMMIT_ID = "e93a26e7707683d349bf5d5c41c5b0ef69b677a9"

import jax
import jax.numpy as jnp

# check how many devices are available
jax.local_device_count()
# Load models & tokenizer
from dalle_mini import DalleBart, DalleBartProcessor
from vqgan_jax.modeling_flax_vqgan import VQModel
from transformers import CLIPProcessor, FlaxCLIPModel

# Load dalle-mini
model, params = DalleBart.from_pretrained(
    DALLE_MODEL, revision=DALLE_COMMIT_ID, dtype=jnp.float16, _do_init=False
)

# Load VQGAN
vqgan, vqgan_params = VQModel.from_pretrained(
    VQGAN_REPO, revision=VQGAN_COMMIT_ID, _do_init=False
)

from flax.jax_utils import replicate

params = replicate(params)
vqgan_params = replicate(vqgan_params)

from functools import partial

# model inference
@partial(jax.pmap, axis_name="batch", static_broadcasted_argnums=(3, 4, 5, 6))
def p_generate(
    tokenized_prompt, key, params, top_k, top_p, temperature, condition_scale
):
    return model.generate(
        **tokenized_prompt,
        prng_key=key,
        params=params,
        top_k=top_k,
        top_p=top_p,
        temperature=temperature,
        condition_scale=condition_scale,
    )


# decode image
@partial(jax.pmap, axis_name="batch")
def p_decode(indices, params):
    return vqgan.decode_code(indices, params=params)

import random

# create a random key
seed = random.randint(0, 2**32 - 1)
key = jax.random.PRNGKey(seed)

from dalle_mini import DalleBartProcessor

processor = DalleBartProcessor.from_pretrained(DALLE_MODEL, revision=DALLE_COMMIT_ID)

import mlflow.transformers

import mlflow

with mlflow.start_run() as run:
    mlflow.transformers.log_model(
        transformers_model=model,
        processor=processor,
        artifact_path="img_model",
    )
```

The resulting error message is 

```
Traceback (most recent call last):
  File "image.py", line 77, in <module>
    mlflow.transformers.log_model(
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 467, in log_model
    return Model.log(
  File "/Users/corey.zumar/mlflow/mlflow/models/model.py", line 547, in log
    flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 261, in save_model
    resolved_task = _get_or_infer_task_type(transformers_model, task)
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 727, in _get_or_infer_task_type
    task = _infer_transformers_task_type(model)
  File "/Users/corey.zumar/mlflow/mlflow/transformers.py", line 750, in _infer_transformers_task_type
    raise MlflowException(
mlflow.exceptions.MlflowException: The provided model type: <class 'dalle_mini.model.modeling.DalleBart'> is not supported. Supported model types are: PreTrainedModel or Pipeline.
```

We should instead say that model must be a pipeline or a dict defining particular keys. The current message is inaccurate.